// @thirdweb-dev/wallets/evm/dist/thirdweb-dev-wallets-evm.esm.js
export { C as CredentialsStorage } from '../../dist/CredentialsStorage-88c8666b.esm.js';
export { a as AbstractClientWallet, D as DEFAULT_DAPP_META } from '../../dist/base-0142d377.esm.js';
export { T as TWConnector, W as WagmiAdapter } from '../../dist/tw-connector-d83550c7.esm.js';
export { PaperWallet } from '../wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.esm.js';
export { CoinbaseWallet } from '../wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.esm.js';
export { LocalWallet, isValidPrivateKey } from '../wallets/local-wallet/dist/thirdweb-dev-wallets-evm-wallets-local-wallet.esm.js';
export { InjectedWallet } from '../wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.esm.js';
export { MetaMaskWallet } from '../wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.esm.js';
export { WalletConnect } from '../wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.esm.js';
export { WalletConnectV1 } from '../wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.esm.js';
export { SafeWallet } from '../wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.esm.js';
export { SmartWallet, getAssociatedAccounts, isAccountIdAvailable } from '../wallets/smart-wallet/dist/thirdweb-dev-wallets-evm-wallets-smart-wallet.esm.js';
export { a as assertWindowEthereum } from '../../dist/assertWindowEthereum-bb2b4544.esm.js';
export { A as AddChainError, a as ChainNotConfiguredError, C as Connector, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from '../../dist/errors-64449f60.esm.js';
export { n as normalizeChainId } from '../../dist/normalizeChainId-dd5a6036.esm.js';
export { SafeSupportedChainsSet } from '../connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.esm.js';
import '../../dist/classPrivateMethodGet-71fe23d8.esm.js';
import '../../dist/classPrivateFieldSet-e25e1cec.esm.js';
import '../../dist/defineProperty-e24c82ea.esm.js';
import '../wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.esm.js';
import 'ethers';
import 'eventemitter3';
import '@thirdweb-dev/chains';
import 'buffer';
import '@safe-global/safe-ethers-adapters';
import '@safe-global/safe-core-sdk';
import '@safe-global/safe-ethers-lib';
import '@thirdweb-dev/sdk';
