// @thirdweb-dev/wallets/dist/AsyncJsonFileStorage-90a596f2.browser.esm.js
import { _ as _defineProperty } from './defineProperty-c8ecdc07.browser.esm.js';
import fsModule from 'fs';
import path from 'path';

const fs = fsModule.promises;
class AsyncJsonFileStorage {
  constructor(filePath) {
    _defineProperty(this, "filePath", void 0);
    this.filePath = path.resolve(filePath);
  }
  async getItem(key) {
    const content = await fs.readFile(this.filePath, {
      encoding: "utf-8"
    });
    if (!content) {
      return null;
    }
    const data = JSON.parse(content);
    return data[key];
  }
  async setItem(key, value) {
    // if the file doesn't exist, create it

    try {
      const content = await fs.readFile(this.filePath, {
        encoding: "utf-8"
      });
      const data = content ? JSON.parse(content) : {};
      data[key] = value;
      await fs.writeFile(this.filePath, JSON.stringify(data));
    } catch {
      await fs.writeFile(this.filePath, JSON.stringify({
        [key]: value
      }));
    }
  }
  async removeItem(key) {
    const content = await fs.readFile(this.filePath, {
      encoding: "utf-8"
    });
    const data = JSON.parse(content);
    delete data[key];
    await fs.writeFile(this.filePath, JSON.stringify(data));
  }
}

export { AsyncJsonFileStorage as A };
